### DISCUSSION TITLE
The FLTK app is not found in task manager and taskbar when window border is set false

### DISCUSSION THREAD

**Discussion by santokalayil** [fltk-rs/fltk-rs#837] [Q&A]
The FLTK app is not found in task manager and taskbar when window border is set false.
I was trying to create a custom title bar after hiding the default title bar (by setting `win.set_border(false)`)
How should I get the app window shown in taskbar, and task manager while keeping default titlebar hidden?
Or is it possible to customize the default titlebar using fltk crate?

![issue](https://user-images.githubusercontent.com/49450970/119601875-ad82a180-be07-11eb-9292-99c7e23e87fe.png)

**Comment by MoAlyousef** [comment]
Borderless is mostly used for popups/dialogs. If you need it for an app window, you'll have to call platform specific code, which for Windows you might want to use the winapi crate. 
Basically you'll need to get the main windows raw_handle() which returns an HWND. Get the GWL_EXSTYLE styles, add 
 WS_EX_APPWINDOW, then adjust. 

Not sure how that's done using the winapi crate. This just wraps the functions and consts we need:
```rust
use fltk::{prelude::*, *};
use std::os::raw::*;

const GWL_EXSTYLE: i32 = -20;
const WS_EX_APPWINDOW: c_ulong = 0x00040000;
extern "C" {
    pub fn GetWindowLongA(wnd: *mut c_void, idx: c_int) -> c_ulong;
    pub fn SetWindowLongA(wnd: *mut c_void, idx: c_int, newval: c_ulong) -> c_long;
}

fn main() {
    let app = app::App::default();
    let mut win = window::Window::default()
        .with_size(400, 300)
        .with_label("Test");
    win.end();
    win.set_border(false);
    win.show();
    unsafe {
        let handle = win.raw_handle();
        let mut style_ex = GetWindowLongA(handle, GWL_EXSTYLE);
        style_ex |= WS_EX_APPWINDOW;
        SetWindowLongA(handle, GWL_EXSTYLE, style_ex);
    }

    app.run().unwrap();
}
```

**Reply by interstellarvoyager** [reply]
i'm having the same issue. will this also work on linux ? specifically debian/ubuntu/mint ?

**Reply by MoAlyousef** [reply]
No unfortunately this code is windows specific. For X11 systems, you can try win.set_xclass("app_name"). Some distros also require a .desktop file for applications which lists the app name and icon location.

**Reply by interstellarvoyager** [reply]
adding further to the misfortune, neither win.set_xclass("app_name") nor using a .desktop file worked. that's all ok for now. i'll just use the default window border frame and continue on. i may look through fltk's source code again to see what i can deduce. thanks for the tips!

**Reply by interstellarvoyager** [reply]
SOLUTION : after fruitlessly trying to find where in-the-hell border(int) was defined, i settled on simply calling .set_border(false) AFTER calling .show() which seems against the recommendation in fltk's doc. hopefully this doesn't break further down the line.

**Comment by santokalayil** [comment]
Thank you. It worked for me. I made a custom window title bar with almost all functionalities. 
But stuck at Event(double click) on the title bar to maximize and restore.. Also minimize did not work.. Is there some solutions for that also?

![image](https://user-images.githubusercontent.com/49450970/119654992-acbd3000-be46-11eb-8430-7a08840c7b53.png)

**Comment by MoAlyousef** [comment]
To minimize you need to call window.iconize(). To detect double clicks, you need to handle Event::Push then check event_clicks():
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let app = app::App::default();
    let mut win = window::Window::default()
        .with_size(400, 300)
        .with_label("Test");
    let mut frame = frame::Frame::new(0, 0, 400, 40, None);
    frame.set_frame(FrameType::FlatBox);
    frame.set_color(Color::Red);
    win.end();
    win.set_border(false);
    win.show();
    frame.handle(move |f, ev| match ev {
        Event::Push => {
            if app::event_clicks() {
                println!("Double clicked!");
                true
            } else {
                false
            }
        },
        _ => false,
    });
    app.run().unwrap();
}
```

**Comment by santokalayil** [comment]
Also is there a direct method to maximize window while the window border is hidden. Currently I am using win.fullscreen(true) for this.. But it actually works as an alternate method only..

**Comment by MoAlyousef** [comment]
No you'll have to resize the window to position 0, 0 and size app::screen_size().

**Comment by santokalayil** [comment]
Thank you very much..

**Comment by wyhinton** [comment]
@santokalayil Any chance you could share the source code for the app showed in your screenshots? I was thinking about making something similar for my [fltk examples](https://github.com/wyhinton/FLTK-RS-Examples) repo, but haven't gotten around to it yet. Or if you'd like you could make a PR.

**Comment by santokalayil** [comment]
@![image](https://user-images.githubusercontent.com/49450970/119763527-534c1400-becd-11eb-8bb1-d7f1f8ec0028.png) hi, I have not completed the titlebar fully.. I actually moved on with whatever I have now for the time being to do the things that are more necessary.. I thought I will come back and do them later. 
The basic Idea is to handle all mouse events on frame and buttons which are used to construct titlebar. so far I have done untill this much
![image](https://user-images.githubusercontent.com/49450970/119763543-5cd57c00-becd-11eb-92c5-c54cced07f8f.png)
Once completed I will send you the code. Or I will send the code here whatever nessessory. Please see a little snippet of my code from here.
`        title_frame.handle(
            move |g, ev| 
            {

                match ev {
                fltk::enums::Event::Enter => {
                    g.parent().unwrap().set_color(Color::from_u32(0x990000));
                    cls_btn_clone.set_color(Color::from_u32(0xaa0000));
                    fltk::app::redraw();
                    true
                }
                fltk::enums::Event::Leave => {
                    g.parent().unwrap().set_color(bg_color);
                    cls_btn_clone.set_color(bg_color);
                    fltk::app::redraw();
                    true
                }
                fltk::enums::Event::Move => {
                    true
                }
                fltk::enums::Event::Released => {
                    fltk::app::redraw();
                    true
                }
                _ => false,
                }
            }
        );
....
`
This is the basic Idea

![image](https://user-images.githubusercontent.com/49450970/119764583-50522300-becf-11eb-8c37-a9dd01434e65.png)

Still not getting the idea, I shall share the code right away as incomplete code

**Comment by santokalayil** [comment]
@wyhinton Customized TitleBar in action can be found in this following youtube video
[https://youtu.be/D0pfY6NUvYM](https://youtu.be/D0pfY6NUvYM)  
![issue2](https://user-images.githubusercontent.com/49450970/119769775-b8f1cd80-bed8-11eb-8b30-1f4524b99d76.png)

**Comment by santokalayil** [comment]
Hi, @wyhinton, The code is available in the [repository](https://github.com/santokalayil/fltk-rust-custom_titlebar). Not completed yet. still uploaded to github.. please take a look at it...

**Comment by santokalayil** [comment]
Hi, @MoAlyousef , 
**2 doubts:**
1. My app quits when esc key is pressed. I don't know why this happens. I have not done anything to detect keyboard presses and not assigned any functions to quit the app. Why this happens?
2. I was trying to add the custom made titlebar the functionality of moving the window by holding and dragging the window's custom title bar. I don't know how to implement that functionality.  To simply the complexity of the project, i was using multiple rs files as modules. So I kept the titlebar as pub struct in a module called titlebar. Please help me with some ideas.

The repository link is here - [https://github.com/santokalayil/fltk-rust-custom_titlebar](https://github.com/santokalayil/fltk-rust-custom_titlebar)

**Comment by MoAlyousef** [comment]
Hi

1- That's FLTK's default behavior, it can be overriden:
```rust
    win.set_callback(move |_| {
            if app::event() == Event::Close {
                app::quit(); // This basically ignores other callback triggers (including hitting esc)
            }
        });
```

2- It should be simple, you'll need to handle Event::Drag. Taking the previous example:
```rust
use fltk::{enums::*, prelude::*, *};

fn main() {
    let app = app::App::default();
    let mut win = window::Window::default()
        .with_size(400, 300)
        .with_label("Test");
    let mut frame = frame::Frame::new(0, 0, 400, 40, None);
    frame.set_frame(FrameType::FlatBox);
    frame.set_color(Color::Red);
    win.end();
    win.set_border(false);
    win.show();
    frame.handle({
        let mut x = 0;
        let mut y = 0;
        move |f, ev| match ev {
            enums::Event::Push => {
                if app::event_clicks() {
                    println!("Double clicked!");
                } else {
                    let coords = app::event_coords();
                    x = coords.0;
                    y = coords.1;
                }
                true
            }
            enums::Event::Drag => {
                f.window()
                    .unwrap()
                    .set_pos(app::event_x_root() - x, app::event_y_root() - y);
                true
            }
            _ => false,
        }
    });
    app.run().unwrap();
}
```

**Comment by santokalayil** [comment]
Thanks, it worked!
